File: contracts/typescript/src/registry.ts
/**
 * Schema registry: mirrors contracts/python/mirage_contracts/registry.py.
 * Reads JSON Schemas bundled under src/schemas/{events,commands,api}/ at
 * runtime (populated by `make generate-contracts` / scripts/generate-contracts,
 * copied verbatim from the repository-root /schemas/ source of truth).
 */
import { readdirSync, readFileSync, existsSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { createRequire } from "node:module";
import path from "node:path";
import type { ValidateFunction } from "ajv";

// Minimal shape of what this module actually calls on an Ajv instance — kept
// narrow deliberately (see below for why these are loaded via createRequire
// rather than a static default import).
interface AjvLike {
  compile(schema: object): ValidateFunction;
}
interface Ajv2020Constructor {
  new (opts?: { allErrors?: boolean; strict?: boolean }): AjvLike;
}

// ajv/ajv-formats ship CJS builds whose default-export interop under
// moduleResolution=NodeNext is ambiguous across TS/Node versions (ajv's
// package.json declares no "exports" map and no "type", so TS's NodeNext
// implied-format detection for its .d.ts disagrees with the ESM `export
// default` syntax that .d.ts actually uses, and types the default binding as
// the whole module namespace instead of the constructable class). Loading
// via createRequire sidesteps that resolution entirely — a plain runtime CJS
// require, typed narrowly against only the surface this file uses.
const require = createRequire(import.meta.url);
const Ajv2020 = require("ajv/dist/2020.js") as unknown as Ajv2020Constructor;
const addFormats = require("ajv-formats") as unknown as (ajv: AjvLike) => void;
import { MalformedSchemaVersionError } from "./errors.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const BUNDLED_SCHEMAS_DIR = path.join(__dirname, "schemas");

const VERSIONED_FILENAME_RE = /^(?<typeName>[a-z][a-z0-9_.]*)\.v(?<major>[0-9]+)\.schema\.json$/;
export const SCHEMA_VERSION_RE = /^(?<major>[0-9]+)\.(?<minor>[0-9]+)$/;

export function parseSchemaVersion(schemaVersion: string): [number, number] {
  const m = SCHEMA_VERSION_RE.exec(schemaVersion);
  if (!m || !m.groups) {
    throw new MalformedSchemaVersionError(schemaVersion);
  }
  return [Number(m.groups.major), Number(m.groups.minor)];
}

function makeAjv() {
  const ajv = new Ajv2020({ allErrors: true, strict: false });
  addFormats(ajv);
  return ajv;
}

export class SchemaRegistry {
  readonly kind: "events" | "commands" | "api";
  private validators = new Map<string, ValidateFunction>(); // key: `${typeName}@${major}`
  private raw = new Map<string, object>();
  private envelopeValidator: ValidateFunction | null = null;
  private ajv = makeAjv();

  constructor(kind: "events" | "commands" | "api", schemasDir: string = BUNDLED_SCHEMAS_DIR) {
    this.kind = kind;
    const dir = path.join(schemasDir, kind);
    if (!existsSync(dir)) return;
    for (const filename of readdirSync(dir).sort()) {
      if (!filename.endsWith(".schema.json")) continue;
      const schema = JSON.parse(readFileSync(path.join(dir, filename), "utf-8"));
      if (filename === "envelope.schema.json") {
        this.envelopeValidator = this.ajv.compile(schema);
        continue;
      }
      const m = VERSIONED_FILENAME_RE.exec(filename);
      if (!m || !m.groups) continue;
      const key = `${m.groups.typeName}@${m.groups.major}`;
      this.validators.set(key, this.ajv.compile(schema));
      this.raw.set(key, schema);
    }
  }

  getEnvelopeValidator(): ValidateFunction {
    if (!this.envelopeValidator) {
      throw new Error(`no envelope.schema.json bundled for kind=${this.kind}`);
    }
    return this.envelopeValidator;
  }

  supportedMajors(typeName: string): Set<number> {
    const out = new Set<number>();
    for (const key of this.validators.keys()) {
      const [name, major] = key.split("@");
      if (name === typeName) out.add(Number(major));
    }
    return out;
  }

  getValidator(typeName: string, major: number): ValidateFunction | undefined {
    return this.validators.get(`${typeName}@${major}`);
  }
}

export const eventsRegistry = new SchemaRegistry("events");
export const commandsRegistry = new SchemaRegistry("commands");

File: contracts/python/mirage_contracts/registry.py
"""Schema registry: maps (kind, type_name) -> {major_version: compiled schema}.

Reads from the schemas/ directory BUNDLED inside this package
(mirage_contracts/schemas/), not from the repository-root /schemas directly —
that bundled copy is produced by `scripts/generate-contracts` (`make
generate-contracts`), so a built/installed mirage_contracts wheel is
self-contained (works inside a Docker image with no repo checkout present).
The repository-root /schemas/ tree remains the single hand-edited source of
truth; `make validate-contracts` fails CI if the bundled copy has drifted
from it (see scripts/generate-contracts and scripts/validate-contracts).

File naming convention read here: `<type_name>.v<major>.schema.json`
(e.g. `agent.heartbeat.v1.schema.json`). The envelope schemas
(`envelope.schema.json`, unversioned filename) are handled separately.
"""
from __future__ import annotations

import json
import re
from pathlib import Path

from jsonschema import Draft202012Validator

from mirage_contracts.errors import MalformedSchemaVersionError

_BUNDLED_SCHEMAS_DIR = Path(__file__).resolve().parent / "schemas"

_VERSIONED_FILENAME_RE = re.compile(r"^(?P<type_name>[a-z][a-z0-9_.]*)\.v(?P<major>[0-9]+)\.schema\.json$")

SCHEMA_VERSION_RE = re.compile(r"^(?P<major>[0-9]+)\.(?P<minor>[0-9]+)$")


def parse_schema_version(schema_version: str) -> tuple[int, int]:
    m = SCHEMA_VERSION_RE.match(schema_version)
    if not m:
        raise MalformedSchemaVersionError(schema_version)
    return int(m.group("major")), int(m.group("minor"))


class SchemaRegistry:
    """Loads and caches JSON Schemas + jsonschema validators for a given kind
    ('events' or 'commands'), keyed by (type_name, major_version).
    """

    def __init__(self, kind: str, schemas_dir: Path | None = None) -> None:
        if kind not in ("events", "commands", "api"):
            raise ValueError(f"unknown kind: {kind!r}")
        self.kind = kind
        self._dir = (schemas_dir or _BUNDLED_SCHEMAS_DIR) / kind
        self._validators: dict[tuple[str, int], Draft202012Validator] = {}
        self._raw: dict[tuple[str, int], dict] = {}
        self._envelope_validator: Draft202012Validator | None = None
        self._load()

    def _load(self) -> None:
        if not self._dir.is_dir():
            return
        for path in sorted(self._dir.glob("*.schema.json")):
            schema = json.loads(path.read_text())
            if path.name == "envelope.schema.json":
                self._envelope_validator = Draft202012Validator(schema)
                continue
            m = _VERSIONED_FILENAME_RE.match(path.name)
            if not m:
                continue  # e.g. unversioned api/error.schema.json — loaded on demand by name
            key = (m.group("type_name"), int(m.group("major")))
            self._validators[key] = Draft202012Validator(schema)
            self._raw[key] = schema

    @property
    def envelope_validator(self) -> Draft202012Validator:
        assert self._envelope_validator is not None, f"no envelope.schema.json bundled for kind={self.kind!r}"
        return self._envelope_validator

    def supported_majors(self, type_name: str) -> set[int]:
        return {major for (name, major) in self._validators if name == type_name}

    def get_validator(self, type_name: str, major: int) -> Draft202012Validator | None:
        return self._validators.get((type_name, major))

    def get_raw_schema(self, type_name: str, major: int) -> dict | None:
        return self._raw.get((type_name, major))

    def known_type_names(self) -> set[str]:
        return {name for (name, _major) in self._validators}


def load_unversioned_schema(kind: str, filename: str, schemas_dir: Path | None = None) -> dict:
    """Load a standalone (non-versioned) schema, e.g. schemas/api/error.schema.json."""
    d = (schemas_dir or _BUNDLED_SCHEMAS_DIR) / kind / filename
    return json.loads(d.read_text())


events_registry = SchemaRegistry("events")
commands_registry = SchemaRegistry("commands")
